File: 2021/haskell/day10.hs
{-# LANGUAGE FlexibleContexts #-}
import Control.Arrow
import Data.Traversable
import Control.Applicative
import Control.Monad.Except
import Data.Functor
import System.Exit
import System.IO
import System.Environment
import Data.Maybe
import Data.List
import Control.Monad.State
import Data.Foldable
import Data.Either

usage :: String -> IO ()
usage progname = putStrLn $ "usage: " ++ progname ++ " <file>"

counterpart :: Char -> Char
counterpart '<' = '>'
counterpart '(' = ')'
counterpart '[' = ']'
counterpart '{' = '}'

-- get a line and output either an autocomplete or a syntax error
checkLine :: String -> Either Char String
checkLine = flip evalState "" . runExceptT . checkLine'
    where checkLine' = traverse_ checkChar >=> const get
          checkChar x
            | x `elem` "<([{" = modify' (counterpart x :)
            | otherwise = do
                -- pop the last opening, if it was empty or it's not the expected one
                -- then say that the character is the problem. Put the rest to make sure it was popped.
                current <- gets uncons
                case current of
                  Nothing -> throwError x
                  Just (c, cs) -> if c /= x then throwError x else put cs


errorScore :: String -> Int
errorScore = mapMaybe score  >>> sum
    where score ')' = Just 3 
          score ']' = Just 57
          score '}' = Just 1197
          score '>' = Just 25137
          score _ = Nothing


autocompleteScore :: String -> Int
autocompleteScore = mapMaybe score >>> foldl (\x a -> x * 5 + a) 0
    where score ')' = Just 1
          score ']' = Just 2
          score '}' = Just 3
          score '>' = Just 4
          score _ = Nothing


completionScore :: [String] -> Int
completionScore = map autocompleteScore >>> sort >>> half
    where half xs = xs !! (length xs `div` 2)

day10 :: String -> (Int, Int)
day10 = lines >>> map checkLine >>> partitionEithers >>> errorScore *** completionScore

main :: IO ()
main = do
    progname <- getProgName
    name <- (getArgs <&> uncons) >>= maybe (usage progname >> exitFailure) return . fmap fst
    (part1, part2) <- readFile name <&> day10
    putStrLn $ "Part 1: " ++ show part1
    putStrLn $ "Part 2: " ++ show part2

File: 2021/golang/day10.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"sort"
)

var SCORES map[rune]uint = map[rune]uint{'>': 25137, '}': 1197, ']': 57, ')': 3}

func checkLine(line string) (rune, []rune) {
	var last []rune
	for _, ch := range line {
		switch ch {
		case '<':
			last = append([]rune{'>'}, last...)
		case '{':
			last = append([]rune{'}'}, last...)
		case '[':
			last = append([]rune{']'}, last...)
		case '(':
			last = append([]rune{')'}, last...)

		case '>':
			fallthrough
		case ']':
			fallthrough
		case '}':
			fallthrough
		case ')':
			if last[0] != ch {
				return ch, last
			}
			last = last[1:]
		}
	}
	return 0, last
}

func completionScore(completion []rune) int {
	score := 0
	for _, ch := range completion {
		score *= 5
		switch ch {
		case ')':
			score += 1
		case ']':
			score += 2
		case '}':
			score += 3
		case '>':
			score += 4
		}
	}
	return score
}

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintf(os.Stderr, "usage: %s <file>\n", os.Args[0])
		os.Exit(1)
	}

	file, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}

	buf := bufio.NewScanner(file)

	var partOne uint = 0
	var partTwo []int

	for buf.Scan() {
		line := buf.Text()
		if errch, ac := checkLine(line); errch != 0 {
			partOne += SCORES[errch]
		} else {
			partTwo = append(partTwo, completionScore(ac))
		}
	}

	if err := buf.Err(); err != nil {
		log.Fatal(err)
	}
	sort.Ints(partTwo)

	fmt.Printf("Part one: %v\n", partOne)
	fmt.Printf("Part two: %v\n", partTwo[len(partTwo)/2])

}
